kvbc/tools/db_restore/db_restore.cpp

```diff
@@ -0,0 +1,122 @@
+      {categorization::kExecutionEventsCategory, categorization::CATEGORY_TYPE::immutable},
+      {categorization::kRequestsRecord, categorization::CATEGORY_TYPE::immutable},
+      {categorization::kExecutionEventGroupDataCategory, categorization::CATEGORY_TYPE::immutable},
+      {categorization::kExecutionEventGroupTagCategory, categorization::CATEGORY_TYPE::immutable},
+      {categorization::kExecutionEventGroupLatestCategory, categorization::CATEGORY_TYPE::versioned_kv},
+      {categorization::kConcordInternalCategoryId, categorization::CATEGORY_TYPE::versioned_kv},
+      {categorization::kConcordReconfigurationCategoryId, categorization::CATEGORY_TYPE::versioned_kv}};
+
+  kv_blockchain_ = std::make_unique<categorization::KeyValueBlockchain>(
+      storage::rocksdb::NativeClient::fromIDBClient(rocksdb_dataset_.dataDBClient), linkStChain, kvbc_categories);
+}
+/**
+ * Optimistic approach for a single pass.
+ * Retrieve blocks from object store in a natural order and add them to rocksdb while comparing their digests.
+ * At the end compare the last block's digest with a one in latest checkpoint descriptor.
+ */
+void DBRestore::restore() {
+  auto [last_os_block_id, last_os_block_digest] = integrity_check_.getLatestsCheckpointDescriptor();
+  auto last_reachable_rocks_blockid = kv_blockchain_->getLastReachableBlockId();
+  Digest last_reachable_rocksdb_block_digest;
+  if (last_reachable_rocks_blockid == 0) {
+    LOG_WARN(logger_, "rocksdb is empty, will restore the whole blockchain");
+  } else {
+    auto buffer =
+        categorization::RawBlock::serialize(kv_blockchain_->getRawBlock(last_reachable_rocks_blockid).value());
+    auto last_reachable_rocks_block = Sliver::copy(reinterpret_cast<const char*>(buffer.data()), buffer.size());
+    last_reachable_rocksdb_block_digest =
+        integrity_check_.computeBlockDigest(last_reachable_rocks_blockid, last_reachable_rocks_block);
+  }
+  LOG_INFO(logger_,
+           "Last reachable block in rocksdb: " << last_reachable_rocks_blockid
+                                               << " digest: " << last_reachable_rocksdb_block_digest.toString());
+
+  Digest expected_parent_digest = last_reachable_rocksdb_block_digest;
+  for (auto block_id = last_reachable_rocks_blockid + 1; block_id <= last_os_block_id; ++block_id) {
+    auto [digest, raw_block] = integrity_check_.getBlock(block_id);
+    Digest parent_block_digest;
+    ConcordAssert(raw_block.data.parent_digest.size() == DIGEST_SIZE);
+    static_assert(sizeof(Digest) == DIGEST_SIZE);
+    memcpy(const_cast<char*>(parent_block_digest.get()), raw_block.data.parent_digest.data(), DIGEST_SIZE);
+    LOG_INFO(logger_,
+             "block: " << block_id << " digest: " << digest.toString()
+                       << ", parent digest: " << parent_block_digest.toString());
+    if (expected_parent_digest != parent_block_digest)
+      throw std::runtime_error("block " + std::to_string(block_id) +
```

Comment:
This is off-topic, but we need to upgrade to c++20 and use std::fmt (or use is as a library in 17).